pkg/controller/keystoneendpoint/keystoneendpoint_controller.go

```diff
@@ -0,0 +1,227 @@
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var log = logf.Log.WithName("controller_keystoneendpoint")
+
+/**
+* USER ACTION REQUIRED: This is a scaffold file intended for the user to modify with their own Controller
+* business logic.  Delete these comments after modifying this file.*
+ */
+
+// Add creates a new KeystoneEndpoint Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileKeystoneEndpoint{client: mgr.GetClient(), scheme: mgr.GetScheme()}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("keystoneendpoint-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource KeystoneEndpoint
+	err = c.Watch(&source.Kind{Type: &keystonev1.KeystoneEndpoint{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	// TODO(user): Modify this to be the types you create that are owned by the primary resource
+	// Watch for changes to secondary resource Pods and requeue the owner KeystoneEndpoint
+	err = c.Watch(&source.Kind{Type: &corev1.Pod{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &keystonev1.KeystoneEndpoint{},
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileKeystoneEndpoint implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileKeystoneEndpoint{}
+
+// ReconcileKeystoneEndpoint reconciles a KeystoneEndpoint object
+type ReconcileKeystoneEndpoint struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a KeystoneEndpoint object and makes changes based on the state read
+// and what is in the KeystoneEndpoint.Spec
+// TODO(user): Modify this Reconcile function to implement your Controller logic.  This example creates
+// a Pod as an example
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileKeystoneEndpoint) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling KeystoneEndpoint")
+
+	// Fetch the KeystoneEndpoint instance
+	instance := &keystonev1.KeystoneEndpoint{}
```

Comment:
I left the same comment on the KeystoneService controller but we'll want to block here to ensure the actual keystone service is bootstrapped before continuing. Doesn't have to be this spot, but somewhere in the controller.

Reply:
Why blocking things instead of leaving it for the another reconciliation attempt? I thought blocking things in reconcile is not recommended.

Reply:
i think we need to decide on the pattern we want here since this will be a scenario we will see in many places. I tend to agree with bogdan that returning a requeue result seems like it would be more of a k8s like pattern. we can even set a delay before the reconcile is attempted again.

could we set something like reconcileAttempts in the Status, and if we exceed a certain amount, then return an error? That, in combination with the reconcile delay would give us what we need to block sufficiently without leaving the reconcile loop blocking for a long period of time.

Reply:
Yes. Result.Requeue is what I was referring too about. That would effectively block the controller from finishing until keystone bootstrapping is completed. I'm already using Result.Requeue in the main keystone controller fwiw, and the pattern is very common across operators